instructure/canvas-ios
# Progress logic integration

refs: MBL-16840
affects: Student
release note: none
test plan: See more below

### This PR integrates the [Progress reporting logic](https://github.com/instructure/canvas-ios/pull/2481) and contains the following changes: 
- Extracted CourseSyncEntry related entities to their own files
- Reworked CourseSyncEntry’s index based selection logic look up to be a custom id based one 
- Refactored `CourseSyncInteractor` for better readability 
- Added a clean up method to `CourseSyncProgressWriterInteractor` so file progress is reseted when a new download starts
- Added `DeleteLocalUseCase`
- Added dedicated `LocalUseCase`, `DeleteLocalUseCase` class definitions for progress related logic
- Added File size to `CourseSyncSelection` and `CourseSyncProgress` screens  
- Renamed `CourseSyncFileFolderInteractor` to `CourseSyncEntryComposerInteractor` to better express its responsibility
- Added the `getEntitiesFromDatabase` function to `ReactiveStore` so we can directly read from `Core Data`   
-  Renamed `XCTAssertSingleOutputEquals` to `XCTAssertCompletableSingleOutputEquals` and added a new `XCTAssertSingleOutputEquals` implementation so we can assert on non-completable publishers as well

### What's not part of this PR 
- Progress UI opens from a temporary kebab menu item and not from the Dashboard Notification.
- Cancellation.
- Retry.

## Testing 
0. You can use my user to test lengthy downloads as I have plenty of uploaded files 
1. Select kebab menu on the Dashboard. Select `Manage Offline Content`. 
2. Select kebab menu on the Dashboard. Select `Offline Content Downloads`. 
3. You should only see courses/tabs/files that were previously selected. 
4. You should see them progressing. 

## Screenshots

<table>
<tr><th>In Progress - Collapsed</th><th>In Progress - Expanded</th></tr>
<tr>
<td><img src="https://github.com/instructure/canvas-ios/assets/110813321/f7f906d4-f772-4be5-8b2a-2bb3b3f7fc64" maxHeight=300></td>
<td><img src="https://github.com/instructure/canvas-ios/assets/110813321/68cf5514-85ad-4bc3-924d-b0b91d91dcfb" maxHeight=300></td>
</tr>
<tr><th>Error state</th><th>CourseSyncSelector file size</th></tr>
<tr>
<td><img src="https://github.com/instructure/canvas-ios/assets/110813321/e4a8ba30-c4e7-44b2-b0ed-e6d67bba01ae" maxHeight=300></td>
<td><img src="https://github.com/instructure/canvas-ios/assets/110813321/c0022792-288f-4133-9951-bcbe54cdefeb" maxHeight=300></td>
</tr>
</table>

## Checklist
